actions/rules/action_state_to_metastate_test.go

```diff
@@ -0,0 +1,433 @@
+		a = action.removeElement(a, 2)
+		expected := []interface{}{0, 1, 3}
+		require.Len(t, a, 3)
+		require.Equal(t, expected, a)
+	})
+}
+
+func (s *ActionStateToMetastateSuite) TestDifference() {
+	s.T().Run("finding differences", func(t *testing.T) {
+		fxt := tf.NewTestFixture(t, s.DB, tf.CreateWorkItemEnvironment())
+		action := ActionStateToMetaState{
+			Db:     s.GormDB,
+			Ctx:    s.Ctx,
+			UserID: &fxt.Identities[0].ID,
+		}
+		a := []interface{}{0, 1, 2, 3, 2}
+		b := []interface{}{2, 3, 5}
+		added, removed := action.difference(a, b)
+		require.Len(t, added, 1)
+		require.Len(t, removed, 2)
+		// wasting plenty more memory here
+		var expectedAdded []interface{}
+		expectedAdded = append(expectedAdded, 5)
+		var expectedRemoved []interface{}
+		expectedRemoved = append(expectedRemoved, 0)
+		expectedRemoved = append(expectedRemoved, 1)
+		require.Equal(t, added, expectedAdded)
+		require.Equal(t, removed, expectedRemoved)
+	})
+}
+
+func (s *ActionStateToMetastateSuite) TestActionExecution() {
+	// Note on the fixture: by default, the created board is attached to the type group
+	// with the same index. The columns on each board are as follows:
+	//   0 "New" "mNew"
+	//   1 "In Progress" "mInprogress"
+	//   2 "Resolved" "mResolved"
+	//   3 "Approved" "mResolved"
+	// All columns are set to the "BidirectionalStateToColumn" rule. The type has
+	// the following states/metastates:
+	//   "new"         "mNew"
+	//   "open"        "mOpen"
+	//   "in progress" "mInprogress"
+	//   "resolved"    "mResolved"
+	//   "closed"      "mClosed"
+	fxtFn := func(t *testing.T) *tf.TestFixture {
+		// this sets up a work item in state "new" with matching metastate "mNew" and the corresponding column 0 ("New")
+		// this represents a single work item in an ordered "new" state.
+		return tf.NewTestFixture(t, s.DB, tf.CreateWorkItemEnvironment(), tf.WorkItemBoards(1), tf.WorkItems(1, func(fxt *tf.TestFixture, idx int) error {
+			fxt.WorkItems[idx].Fields[workitem.SystemState] = workitem.SystemStateNew
+			fxt.WorkItems[idx].Fields[workitem.SystemMetaState] = "mNew"
+			fxt.WorkItems[idx].Fields[workitem.SystemBoardcolumns] = []string{
+				fxt.WorkItemBoards[idx].Columns[0].ID.String(),
+			}
+			return nil
+		}))
+	}
+	// test the fixture creation.
+	fxt := fxtFn(s.T())
+	require.NotNil(s.T(), fxt)
+	require.Len(s.T(), fxt.WorkItems, 1)
+	require.Equal(s.T(), "mNew", fxt.WorkItems[0].Fields[workitem.SystemMetaState])
+	require.Equal(s.T(), workitem.SystemStateNew, fxt.WorkItems[0].Fields[workitem.SystemState])
+	require.Equal(s.T(), fxt.WorkItemBoards[0].Columns[0].ID.String(), fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns].([]interface{})[0])
+
+	s.T().Run("updating the state for an existing work item", func(t *testing.T) {
+		fxt := fxtFn(t)
+		// set the state to "in progress" and create a changeset.
+		fxt.WorkItems[0].Fields[workitem.SystemState] = workitem.SystemStateInProgress
+		contextChanges := change.Set{
```

Comment:
Please use `workitem.ChangeSet(newWorkItem)`

Reply:
This requires a copy of the fxt WorkItem. As this is then repeated from the actions whitebox tests, I looked into `github.com/jinzhu/copier`. Turns out this does not support maps. Other libs would introduce a lot of additional complexity, not to talk about additional deps.

And as the action execution does not even deal with the "old" WI instance (it is *only* required to create the changeset) and the changeset creation is already tested in the actions whitebox test, I'll suggest we leave it as it is, creating the changeset from scratch.

Reply:
>This requires a copy of the fxt WorkItem. As this is then repeated from the actions whitebox tests, I looked into github.com/jinzhu/copier. Turns out this does not support maps. Other libs would introduce a lot of additional complexity, not to talk about additional deps.

Thank you for looking into it. I am fine with creating the `change.Set{}` struct.